Octokit/Models/Request/BranchProtectionUpdate.cs

```diff
@@ -304,28 +293,146 @@ internal string DebuggerDisplay
+        /// <param name="requireCodeOwnerReviews">Blocks merge until code owners have reviewed.</param>
+        public BranchProtectionRequiredReviewsUpdate(BranchProtectionRequiredReviewsDismissalRestrictionsUpdate dismissalRestrictions, bool dismissStaleReviews, bool requireCodeOwnerReviews)
         {
-            IncludeAdmins = includeAdmins;
+            Ensure.ArgumentNotNull(dismissalRestrictions, nameof(dismissalRestrictions));
+
+            DismissalRestrictions = dismissalRestrictions;
+            DismissStaleReviews = dismissStaleReviews;
+            RequireCodeOwnerReviews = requireCodeOwnerReviews;
         }
 
         /// <summary>
-        /// Enforce pull request reviews for repository administrators
+        /// Specify which users and teams can dismiss pull request reviews.
+        /// </summary>
+        public BranchProtectionRequiredReviewsDismissalRestrictionsUpdate DismissalRestrictions { get; protected set; }
+
+        /// <summary>
+        /// Dismiss approved reviews automatically when a new commit is pushed.
+        /// </summary>
+        public bool DismissStaleReviews { get; protected set; }
+
+        /// <summary>
+        /// Blocks merge until code owners have reviewed.
         /// </summary>
-        public bool IncludeAdmins { get; protected set; }
+        public bool RequireCodeOwnerReviews { get; protected set; }
 
         internal string DebuggerDisplay
         {
             get
             {
-                return string.Format(CultureInfo.InvariantCulture, "IncludeAdmins: {0}", IncludeAdmins);
+                return string.Format(CultureInfo.InvariantCulture, "DismissalRestrictions: {0} DismissStaleReviews: {1} RequireCodeOwnerReviews: {2}",
+                    DismissalRestrictions?.DebuggerDisplay ?? "disabled",
+                    DismissStaleReviews,
+                    RequireCodeOwnerReviews);
             }
         }
+    }
+
+    /// <summary>
+    /// Specifies whether review dismissal for the protected branch will be restricted to Admins, specified Teams/Users or unrestricted
+    /// </summary>
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public class BranchProtectionRequiredReviewsDismissalRestrictionsUpdate
+    {
+        /// <summary>
+        /// EnableSpecify only administrators are allowed to dismiss reviews
+        /// </summary>
+        /// <param name="enabled">True to restrict review dismissal to Admins, false to disable restriction</param>
+        public BranchProtectionRequiredReviewsDismissalRestrictionsUpdate(bool enabled)
+        {
+            if (enabled)
+            {
+                Teams = new BranchProtectionTeamCollection();
+                Users = new BranchProtectionUserCollection();
+            }
+            else
+            {
+                // API requires an empty array to be passed, to disable these checks
+                Teams = null;
```

Comment:
Is there a way to make it clearer what is happening here?
as a newcomer to the codebase might be confused by `null` being equated with an empty array

Reply:
Yep I agree, this is not great but it's the best I could come up with 😡 

Im not sure why this API endpoint is different to the other branch protection ones!  Normally to "disable" a protection setting, you can send a `null` for the object but in this case to disable the item, you have to send an empty bodied member.

So basically to be disabled, as per the following screen shot, we have to send an instance of the class but with no members (ie the Teams/Users array are `null`)
![image](https://user-images.githubusercontent.com/5425163/29747619-ce70a9b4-8b43-11e7-9efd-0b9310f0cc17.png)

To enable the dismiss restriction and have only admins (as per next screenshot) you send empty Teams/Users arrays.
![image](https://user-images.githubusercontent.com/5425163/29747625-ebde1964-8b43-11e7-8d06-41573fdfc2e4.png)

To specify teams/users you populate them
![image](https://user-images.githubusercontent.com/5425163/29747628-f8c630d0-8b43-11e7-98ab-816bb2b90970.png)



So given that we need to somehow have either the empty bodied class or empty array or populated arrays, what i came up with was the `ctor` thqat takes a `bool` for whether the restriction is enabled (admin only, thus empty team/user list)... or disabled (`null` arrays, so the class has an empty body).

I tried to explain the usage in the xmldoc comments which users hopefully see in intellisense.

If you have any better ideas for implementation let me know!

Reply:
From the perspective of an external user, I think what you have is a nice solution.
My worry is more on someone coming along later trying to contribute, and not understanding the code comment
Maybe something like `API requires an object with empty members to be passed in order to disable these checks`?

Reply:
Ah right, sorry if misunderstood. Yes that code comment is poorly worded, I will update thanks!